src/java/org/apache/cassandra/db/streaming/CassandraOutgoingFile.java

```diff
@@ -114,13 +155,51 @@ public void write(StreamSession session, DataOutputStreamPlus out, int version)
         CassandraStreamHeader.serializer.serialize(header, out, version);
         out.flush();
 
-        CassandraStreamWriter writer = header.compressionInfo == null ?
-                                       new CassandraStreamWriter(sstable, header.sections, session) :
-                                       new CompressedCassandraStreamWriter(sstable, header.sections,
-                                                                           header.compressionInfo, session);
+        IStreamWriter writer;
+        if (shouldStreamFullSSTable())
+        {
+            writer = new CassandraBlockStreamWriter(sstable, session, components);
+        }
+        else
+        {
+            writer = (header.compressionInfo == null) ?
+                     new CassandraStreamWriter(sstable, header.sections, session) :
+                     new CompressedCassandraStreamWriter(sstable, header.sections,
+                                                         header.compressionInfo, session);
+        }
         writer.write(out);
     }
 
+    @VisibleForTesting
+    public boolean shouldStreamFullSSTable()
+    {
+        return isFullSSTableTransfersEnabled && isFullyContained;
+    }
+
+    @VisibleForTesting
+    public boolean fullyContainedIn(List<Range<Token>> normalizedRanges, SSTableReader sstable)
+    {
+        if (normalizedRanges == null)
+            return false;
+
+        RangeOwnHelper rangeOwnHelper = new RangeOwnHelper(normalizedRanges);
+        try (KeyIterator iter = new KeyIterator(sstable.descriptor, sstable.metadata()))
+        {
+            while (iter.hasNext())
+            {
+                DecoratedKey key = iter.next();
+                try
+                {
+                    rangeOwnHelper.check(key);
+                } catch(RuntimeException e)
```

Comment:
Catching `RuntimeException` is really not the way we should be using `RangeOwnHelper` here. Can you refactor `RangeOwnHelper` to introduce a method that would return a `boolean` instead?

Reply:
On a more general note, this is potentially quite an expensive thing to do, especially for big sstables with skinny partitions, and in some cases this will introduce a performance regression.

The whole optimisation is realistically only useful for bootrstrap, decom, and rebuild, with LCS (which is still plenty useful and impactful and worth having). But it wouldn't normally kick in for regular repairs because of the full-cover requirement, and it won't normally kick in for STCS until CASSANDRA-10540 (range aware compaction) is implemented. In those cases having to read through the whole primary index is a perf regression that we shouldn't allow to happen.

The easiest way to avoid it would be to store sstable's effective token ranges in sstable metadata in relation to the node's ranges, making this check essentially free. Otherwise we should probably disable complete sstable streaming for STCS tables, at least until CASSANDRA-10540 is implemented. That however wouldn't address the regression to regular streaming, so keeping ranges in the metadata would be my preferred way to go.

Reply:
I mistakenly thought this index was a sampled index not a full index. Requiring a comparison of every partition key in every sstable for the entire data set seems like a big regression for some use cases. 

I was trying and failing to find the reasoning for why we switched to this.

Reply:
@iamaleksey thank you for the useful feedback. I did discuss this with @krummas and I believe while there was a room for improvement, the thinking back then was that the benefits would outweigh the cost. I looked through the codebase and this was the best way to definitely verify range containment as I was going for correctness. That said, what you suggest is obviously better. I am concerned about scope creep in this PR. Would it be ok if we address it as part of a separate PR?

It would also be useful, if we could design the effective range computation and storage in the metadata. I am not sure what sort of gotchas I might run into.

Reply:
@dineshjoshi Duplicating a reply here: I think it's fine to not block this PR on the improvements, but disable zero copy for STCS for now, and commit that way. Once we optimise the containment check - ideally before 4.0 - enable for STCS as well.

Encoding information about the effective contained ranges is one way. Alternative way would be to invert the ranges we are checking for, and look for presence of keys in the gaps - but relying on summary and index to efficiently search rather than linearly scan. But again - let's temporarily disable zero copy for STCS for now, to not delay the ticket beyond necessary.

Reply:
I will disable Zero Copy streaming for STCS and I've created [CASSANDRA-14586](https://issues.apache.org/jira/browse/CASSANDRA-14586) to further discuss fixes.